docs/Versioning.md

```diff
@@ -55,11 +55,142 @@ By way of example, the `ModelProto.ir_version` MUST be present in every model.
 
 ISSUE: decide and document how we want to handle changes to the format that are forward compatible but will cause the generated code to introduce a breaking change (e.g., changing the data type of a field from int64 to int32, int64 to double or bytes to string).  I would recommend that we never do it after we hit 1.0.0 - prior to that we should be fairly liberal with changes to size (e.g., `int32<>int16`, `float<>double`) and conservative with changes to value space (e.g., `integral<>floatingpoint`, `string<>int`, `scalar<>message`)
 
-ISSUE: define type compatiblity rules either here or under model versioning - probably here
+ISSUE: define type compatibility rules either here or under model versioning - probably here
 
 ## Operator versioning
 
-ISSUE: sort out the design and then document.
+ONNX's operator versioning has semantic meaning: the ONNX operator
+version associated with a model controls the set of ONNX operators it
+may use, and the semantic meaning of these operators.  Operator
+versioning permits ONNX to evolve the semantics of the operators it
+specifies in BC-breaking ways, while preserving backwards-compatibility
+with older exported models (which pin the meaning of operators by
+recording the operator version they target).
+
+The ONNX operator version used by a model is recorded in
+`ModelProto.op_version` and is a simple integer (i.e., it does *not*
+use semantic versioning.)  Each operator version of ONNX is associated
+with some machine-readable metadata which specifies the implementation
+for every operator that is in scope.  When ONNX makes BC-breaking
+changes to the semantics of operators (e.g., by changing the meaning
+of an operator or removing an operator), it increments the ONNX
+operator version.  Additions of new operators do NOT increment the
+ONNX operator version (as they are not BC-breaking changes), nor
+do additions of new optional attributes.
+
+When understanding the semantics of an operator version, it is helpful
+to distinguish an *operator implementation* from an operator name
+(`op_type`).  An operator implementation specifies the actual semantics
+of an operator (its type, the mathematical computation it performs on
+its arguments, etc.); an operator name is simply what is recorded in an
+exported model.  Operator implementations are *immutable*; they never
+change across ONNX operator versions.  For example, if ONNX wants to
+introduce a breaking change to the `Conv` operator, it must create a
+new operator implementation describing the new desired behavior, and
+then remap the `Conv` name to this new implementation in the next ONNX
+operator version.  In this way, the ONNX operator version is a
+*namespace* mechanism on top of an immutable store of operator
+implementations; each operator version defines a mapping of operator
```

Comment:
My understanding of this design is, the operator implementation store will never be cleaned up. 

One personal thinking please. I'm in favor of having the capability to clean up operator implementation store. The operator implementation store is part of onnx standard.  We may want to clean up the standard when we feel too old versions do not make sense and we want to enforce an upgrade of vendors' components (importers/exporters). Without the clean up capability, it's always up to vendors to decide the range of operator implementation store they'll support. Does this make sense please?

Reply:
The proposal, as I understand it, is compatible with what you want. The point is that the operator version standard doesn't *mandate* that an ONNX importer actually have implementations for all operator implementations in the store; they can do whatever they want. We can give guidelines about what we recommend they implement fully to make this clearer.